### PR TITLE
Feat/sof 7249

### PR DESCRIPTION
**Pull request by seankwarren**

### CODE REVIEW

File: src/application.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+import { ApplicationConfig, ApplicationData } from "./types";
+
+type ApplicationBase = InstanceType<typeof NamedDefaultableHashedInMemoryEntity>;
+
+// @ts-ignore
+export function ApplicationMixin<
+    T extends Constructor<ApplicationBase> = Constructor<ApplicationBase>
+>(superclass: T) {
+    return class AdeApplication extends superclass {
+        static Executable = Executable;
```
Comment (k0stik) [inline src/application.ts:19]: The error above and below happens because of `static Flavor = Flavor` and `static Executable = Executable`. 
Is there a way to redesign this?

Comment (seankwarren) [reply src/application.ts]: The only workaround I was able to find that works, is by restricting these Entities to classes rather than mixins. This means they cannot be mixed into another class while maintaining type safety, but can be much more easily attached to other classes similar to how `Executable` class is attached to `Application` class as a static attribute. It results in much more reasonable `*.d.ts` file sizes, and doesn't require maintaining class and interface separately. They should be overridable in `web-app` as long as each type is castable to the inferred type in the base Executable: 
example: `flavors` getter expects to return an array. This means the base `Executable` definition should ensure that the type of each method is generic enough for our needs to support overriding. 
```typescript
export class WebappExecutable extends Executable {
    get flavors() {
        return [];
    }
}
export class WebappExecutableApplication extends Application {
    static Executable = WebappExecutable;
}
```

TLDR; 
- Use classes instead of mixins for Application, Executable, Flavor, Template, will need to reconsider ContextProviders since they will prefer mixin functionality.
- Maintain base class interface as part of base class definition, overriding inferred types where necessary to be generic enough.

File: src/application.ts [hunk 2]
```diff
@@ -37,34 +55,34 @@ export class Application extends NamedDefaultableInMemoryEntity {
...
     }
 
     static getUniqueAvailableNames() {
         return allApplications;
     }
 
-    getExecutableByName(name = null) {
-        return new this.constructor.Executable(
+    getExecutableByName(name?: string) {
+        return new Application.Executable(
```
Comment (timurbazhirov) [inline src/application.ts:77]: Did you check that without `this.constructor` re-defining Executable in webapp startup will still work correctly??

File: src/application.ts (hunk 1 above)
Comment (k0stik) [reply src/application.ts:20]: The `application.d.ts` file still seems very large. I will probably create a task for myself to investigate possible solutions for future work. Otherwise code looks good

File: src/application.ts (hunk 2 above)
Comment (seankwarren) [reply src/application.ts]: Seems to work as expected
